src/main/scala/org/constellation/snapshotstreaming/FileBasedLastIncrementalGlobalSnapshotStorage.scala

```diff
@@ -0,0 +1,121 @@
+package org.constellation.snapshotstreaming
+
+import java.nio.file.NoSuchFileException
+
+import cats.effect.Async
+import cats.effect.std.Semaphore
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.either._
+import cats.syntax.eq._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+import cats.{Applicative, MonadThrow}
+
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema._
+import org.tessellation.schema.height.Height
+import org.tessellation.sdk.domain.snapshot.Validator.isNextSnapshot
+import org.tessellation.sdk.domain.snapshot.storage.LastSnapshotStorage
+import org.tessellation.security.Hashed
+
+import fs2.io.file._
+import fs2.{Stream, text}
+import io.circe.Codec
+import io.circe.generic.semiauto.deriveCodec
+import io.circe.parser.decode
+import io.circe.syntax._
+
+object FileBasedLastIncrementalGlobalSnapshotStorage {
+
+  def make[F[_]: Async: Files: KryoSerializer](
+    path: Path
+  ): F[LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo]] =
+    Semaphore[F](1)
+      .map(make(path, _))
+
+  private def make[F[_]: Async: Files: KryoSerializer](
+    path: Path,
+    semaphore: Semaphore[F]
+  ): LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo] =
+    new LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo] {
+
+      implicit val codec: Codec[Hashed[GlobalIncrementalSnapshot]] = deriveCodec[Hashed[GlobalIncrementalSnapshot]]
+      implicit val snapshotWithInfoCodec: Codec[SnapshotWithState] = deriveCodec[SnapshotWithState]
+
+      def set(snapshot: Hashed[GlobalIncrementalSnapshot], state: GlobalSnapshotInfo): F[Unit] =
+        semaphore.permit.use { _ =>
+          validateStateProof(snapshot, state) >>
+            get.flatMap {
+              case Some(last) if isNextSnapshot(last, snapshot.signed.value) =>
+                Stream
+                  .emit(SnapshotWithState(snapshot, state).asJson.spaces2)
```

Comment:
I used `spaces2` when I first implemented the file based storage, to have the file readable without any additional tools, but we could also write it down with `noSpaces` and just pipe through `jq` if readable output is needed. Up to you.